node_modules/oae-principals/lib/internal/dao.js

```diff
@@ -285,6 +289,107 @@ var deletePrincipal = module.exports.deletePrincipal = function(principalId, cal
 };
 
 /**
+ * Marks a principal as being deleted. Currently only groups are supported
+ *
+ * @param  {String}     principalId     The id of the principal to mark as deleted
+ * @param  {Function}   callback        Standard callback function
+ * @param  {Object}     callback.err    An error that occurred, if any
+ */
+var deletePrincipal2 = module.exports.deletePrincipal2 = function(ctx, user, callback) {
+    var deleted = Date.now();
+
+    // Determine if the principal was already deleted in the authz index before we set them and flip
+    // the principals deleted flag
+    AuthzDelete.isDeleted([user.id], function(err, wasDeleted) {
+        if (err) {
+            return callback(err);
+        }
+
+        // Get the login to delete the user form the table AuthenticationLoginId
+        require('oae-authentication').getUserLoginIds(ctx, user.id, function(err, login){
+            if (err) {
+                return callback(err);
+            }
+
+            // Set (or re-Set) the principal as deleted in the authz index
+            AuthzDelete.setDeleted(user.id, function(err) {
+                if (err) {
+                    return callback(err);
+                }
+
+                // Delete user from table Principals
+                Cassandra.runQuery('DELETE FROM "Principals" where "principalId" = ?', [user.id], function(err, rows) {
+                    if (err) {
+                        return callback(err);
+                    }
+
+                    // Delete user from table PrincipalsByEmail
+                    Cassandra.runQuery('DELETE FROM "PrincipalsByEmail" where "email" = ? AND "principalId" = ?', [user.email, user.id], function(err, rows) {
+                        if (err) {
+                            return callback(err);
+                        }
+                    
+                        var loginId = new LoginId(user.tenant.alias, AuthenticationConstants.providers.LOCAL, login.local);
+
+                        // Delete user from table AuthenticationUserLoginId
+                        Cassandra.runQuery('DELETE FROM "AuthenticationUserLoginId" WHERE "userId" = ?', [user.id], function(err, rows) {
+                            if (err) {
+                                return callback(err);
+                            }
+
+                            // Delete user from table AuthenticationLoginId
+                            if(loginId.externalId !== undefined){
+                                Cassandra.runQuery('DELETE FROM "AuthenticationLoginId" WHERE "loginId" = ?', [_flattenLoginId(loginId)], function(err, rows) {
```

Comment:
There are still more tables that the user will need to be deleted from to stop weird behaviour (see #1422 for an example). Ones that come to mind include `AuthzRoles`, possibly other Authz or Authentication tables depending on what their tenant's authentication method is... Maybe do a search for all the tables that use either `principalId` or `userId` to get the full list? This is one of the downsides of a no-sql database.